File: Go刷题/岛屿数量.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:  
            return 0  
    
        rows, cols = len(grid), len(grid[0])  
        count = 0  
    
        def dfs(r, c):  
            if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] != '1':  
                return  
            grid[r][c] = '#'  # 标记为已访问  
            dfs(r - 1, c)  # 上  
            dfs(r + 1, c)  # 下  
            dfs(r, c - 1)  # 左  
            dfs(r, c + 1)  # 右  
    
        for i in range(rows):  
            for j in range(cols):  
                if grid[i][j] == '1':  
                    count += 1  
                    dfs(i, j)  
    
        return count 
            

File: Go刷题/岛屿数量.go
func numIslands(grid [][]byte) int {
    res := 0
    m := len(grid)
    n := len(grid[0])
    for row := 0 ; row < m ; row ++ {
        for col := 0; col < n ;col ++ {
            if grid[row][col] == '1' {
                dfs(grid,row,col)
                res ++
            }
        }
    }

    return res 

}

func dfs(grid [][]byte, r int, c int) {
    m := len(grid)
    n := len(grid[0])

    if r<0 || r >= m || c < 0 ||c >= n || grid[r][c] != '1' {
        return
    }
    grid[r][c] = '#' //标记为已访问
    dfs(grid, r-1,c)
    dfs(grid,r,c-1)
    dfs(grid,r+1,c)
    dfs(grid, r,c+1)
}